Pages & Widgets

PoojaB26 opened the pull request:
### Description
Pages & Widgets doc
Linear ticket and [magic word](https://linear.app/docs/github#link-prs) Fixes DEVR-311 & Fixes DEVR-310

## Type of change
- [ ] Typo fix 
- [x] New feature 
- [ ] Removed outdated references

leighajarett commented on docs/resources/widgets/overview/overview.md:
```diff
@@ -0,0 +1,104 @@
+title: Overview
+---
+# 
+## What's a widget?
+
```
I think we should specifically start by talking about how in Flutter everything is a widget (pages, atomic elements, layouts, etc) -- but in FlutterFlow thats not really the case, pages arent widgets in FlutterFlow but they are in the generated code. I talk a bit about this in my Flutter connection talk

leighajarett replied:
Basically I think we need to talk about pages more here 

Also same comment as before, it feels text heavy. I would recommend making a visual in a slide deck (or using some from the Flutter connection talk) to illustrate your point

PoojaB26 replied:
> but in FlutterFlow thats not really the case, pages arent widgets in FlutterFlow but they are in the generated code

I want to mention this in the generated code page where we show the differences between FlutterFlow's approach and Flutter. I feel it'll be confusing to non Flutter devs to start with "everything in Flutter is a widget but in FlutterFlow its a different approach" because then we are assuming that they are all coming with Flutter understanding. So instead I propose, we just talk about pages as just another entity in FlutterFlow and in generated code page, we talk about the underlying concepts. 

Also there is a pages.md page where we talk about pages, won't that be enough? Also to demo the entire overview of all the entities, maybe that would be covered in Pinkesh's resource hierarchy I think.

leighajarett reviewed (changes requested):
This is looking really great @PoojaB26!

Left a few comments, but generally my feedback falls into a few categories:

- Unclear what the sequence of this content should be - if concept comes first then we shouldn't use terms we ahven't really introduced yet (i.e. widget)

-Overview doesn't really cover pages, I think we need to tie this in more, or separate pages out of the widgets section

-Lets use code block when referring to widget names (nit)

-Most important: need more visuals. There's a lot of text here. I think we need some visual diagrams to illustrate your points more. You gave lots of good examples, so lets show those examples by creating some diagram in a google slide and taking a screenshot of it or something like that! I would recommend starting with the slides I used for Flutter Connection and tweaking to the examples you're using

PoojaB26 commented on docs/resources/widgets/Concepts/concepts.md:
```diff
@@ -0,0 +1,42 @@
+#
+When building widgets and pages in FlutterFlow, it's important to understand what's
+happening behind the scenes. Let's explore some key concepts that are beneficial
+to know.
```
Yes, makes sense. Will move it around!

PoojaB26 commented on docs/resources/widgets/Concepts/concepts.md:
```diff
@@ -0,0 +1,42 @@
+When building widgets and pages in FlutterFlow, it's important to understand what's
+happening behind the scenes. Let's explore some key concepts that are beneficial
+to know.
+## Classes vs Instances
```
Added an example section at the end! Let me know how that looks

PoojaB26 commented:
> Unclear what the sequence of this content should be - if concept comes first then we shouldn't use terms we ahven't really introduced yet (i.e. widget)

Agreed, changed! 

> -Overview doesn't really cover pages, I think we need to tie this in more, or separate pages out of the widgets section

Pages is covered separately in pages.md. 
Hierarchy will probably look like this 

Widgets ( = widgets.md)
-- Concepts
-- Pages 
-- Built In Widgets 
-- Components 
-- Custom Widgets 

> -Most important: need more visuals. There's a lot of text here. I think we need some visual diagrams to illustrate your points more. You gave lots of good examples, so lets show those examples by creating some diagram in a google slide and taking a screenshot of it or something like that! I would recommend starting with the slides I used for Flutter Connection and tweaking to the examples you're using

I did add images at places but there were some rendering issues there, but yes, I agree more visuals would be great, so went ahead and added more images, you can go ahead and check them now! 
Also just fyi, instead of Slides, we can create the illustrations in [Figma here](https://www.figma.com/file/X4dJZ2i8C6kpPrcnHmjBkq/Ecommerce-Demo-App?type=design&node-id=534-26&mode=design&t=88uJzDJTWUlIgJ7z-0) so we won't have to take screenshots and can create high quality illustrations. Just letting you know so you can also use the same and all the images can remain in one place for modification later. 🙌🏽 @leighajarett

PoojaB26 commented:
@leighajarett This PR is ready for review, also added some missing sections in Pages How to. 
@pinkeshmars Also have a look and let me know if any missing sections here

pinkeshmars reviewed (commented):
Seems like there is some issue while reviewing this PR in stackblitz! It just shows blank screen.

Anyways, I think we could add the widget configuration page just like this one (docs/resources/widgets/pages/pages-configuration.md)

leighajarett commented:
@PoojaB26 I am not able to build the docs site with the current version of the branch. Widgets.md wont render, can you please take a look?

leighajarett commented on docs/resources/widgets/widgets.md:
```diff
@@ -0,0 +1,102 @@
+---
+title: Overview
+---
+# 
```
I think we need an h1 in here

leighajarett commented on docs/resources/widgets/widgets.md:
```diff
@@ -0,0 +1,102 @@
+title: Overview
+---
+# 
+## What's a widget?
```
I think we should have a page in the left hand side bar for this page (right now, I have to click on the widget section to get to this page) to be consistent with the projects section. It wasn't really clear to me this existed from the left hand side panel. Perhaps we call it "What's a widget?"

leighajarett commented on docs/resources/widgets/widgets.md:
```diff
@@ -0,0 +1,102 @@
+how other widgets are arranged and displayed on the screen. Common layout
+elements include:
+
+**Row**: Arranges its child widgets horizontally.
+
```
I think these should all be bullets. The page design looks a bit weird with them each as their own paragraph

leighajarett commented on docs/resources/widgets/widgets.md:
```diff
@@ -0,0 +1,102 @@
+**Container**: Provides a box model for a single child widget, with optional
+padding, margins, borders, and background color.
+
+_Check out your builder for the entire list_
```
"the builder" not your builder

leighajarett commented on docs/resources/widgets/widgets.md:
```diff
@@ -0,0 +1,102 @@
+
+### Page Elements
+
+In FlutterFlow, the Page Elements category comprises widgets like AppBar, FAB,
```
We need a consistent way of referring to widgets here. Italics, code, bold -- I don't really care but we should decide one and be consistent.

leighajarett commented on docs/resources/widgets/widgets.md:
```diff
@@ -0,0 +1,102 @@
+and interaction. FlutterFlow leverages a similar concept by providing a visual
+editor where you can create and configure these widgets.
+
+![everything-widget.png](imgs%2Feverything-widget.png)
+
```
This image seems pretty big, I think we can make it smaller

leighajarett commented on docs/resources/widgets/widgets.md:
```diff
@@ -0,0 +1,102 @@
+- Page Elements
+- Form Elements
+
+### Layout Elements
```
For each of these categories, it would be nice to have a table - with an example of the widget in use / image of the example

leighajarett replied:
For example, one row would be for "column widget" - we would have the name / description in one column, and make some visual diagram of a row being used on a page. 

I think we just need some more visuals here, feels too text heavy

leighajarett commented on docs/resources/widgets/widgets.md:
```diff
@@ -0,0 +1,102 @@
+The widget tree in FlutterFlow, organizes the relationships between all the
+widgets in your app's user interface. Each widget can contain other widgets
+inside it, known as "children," creating a nested
+structure. For example, you might have a main screen widget that contains a
```
I'm not sure what you mean here by "main screen widget"

leighajarett commented on docs/resources/widgets/widgets.md:
```diff
@@ -0,0 +1,102 @@
+
+## Widget Tree
+
+The widget tree in FlutterFlow, organizes the relationships between all the
+widgets in your app's user interface. Each widget can contain other widgets
+inside it, known as "children," creating a nested
+structure. For example, you might have a main screen widget that contains a
+column widget, and inside that column, there are text widgets, button widgets,
+and so on.
```
We should describe what's actually being shown in the visual below instead of describing some example that we don't see here

leighajarett commented on docs/resources/widgets/widgets.md:
```diff
@@ -0,0 +1,102 @@
+structure. For example, you might have a main screen widget that contains a
+column widget, and inside that column, there are text widgets, button widgets,
+and so on.
+
+![widget-tree.png](imgs%2Fwidget-tree.png)
```
For all images we need alt text for accessibility purposes

leighajarett commented on docs/resources/widgets/Concepts/key-concepts.md:
```diff
@@ -0,0 +1,61 @@
+sidebar_position: 1
+---
+
+#
```
All pages need an h1

leighajarett commented on docs/resources/widgets/Concepts/key-concepts.md:
```diff
@@ -0,0 +1,61 @@
+is called `topSellingProductCard` for use in the Top Selling section. Similarly,
+in the `CategoryProductListPage`, its instance is called `categoryProductCard`.
+
+![Class-Instance.png](..%2Fimgs%2FClass-Instance.png)
```
I'm a bit lost with this visual - can you make it a bit more clear? I think we need to see more of the UI to orient ourselves

leighajarett commented on docs/resources/widgets/Concepts/composing-widgets.md:
```diff
@@ -0,0 +1,136 @@
+sidebar_position: 2
+---
+
+# Composing widgets
```
I feel like we're repeating a lot of the information from the "widgets.md" file here. I don't think we need to list row, column, etc in both files. I would just keep that information here?

leighajarett commented on docs/resources/widgets/Concepts/composing-widgets.md:
```diff
@@ -0,0 +1,136 @@
+
+![row-col-stack.png](..%2Fimgs%2Frow-col-stack.png)
+
+Use a **Row** for side-by-side layout.
+
+Use a **Column** for a top-to-bottom layout.
+
+Use a **Stack** for overlapping elements.
```
Things like this should be in a bulleted list

leighajarett commented on docs/resources/widgets/Concepts/composing-widgets.md:
```diff
@@ -0,0 +1,136 @@
+---
+title: Composing Widgets
+sidebar_position: 2
```
I feel like this page looks a bit weird. The images take up a lot of space and theres a lot of blank space. I think it makes sense to put this stuff into a tabular format and make the images much smaller so it's easier to view

leighajarett replied:
Like I would have a table with row, stack, column each as their own row and the little images you put as a column in the table

leighajarett commented on docs/resources/widgets/pages/pages.md:
```diff
@@ -0,0 +1,142 @@
+---
+title: Pages
```
Same comment as widgets, I think we need a file nested under the pages section called "Whats a page?" or something. It wasn't clear to me that clicking on Page would bring me to an actual page

PoojaB26 replied:
Yep, corrected!

leighajarett commented on docs/resources/widgets/pages/pages.md:
```diff
@@ -0,0 +1,142 @@
+
+Here's how pages work in FlutterFlow:
+
+#### Structure and Layout
```
We probably want some link back to the docs where we described these

leighajarett commented on docs/resources/widgets/pages/pages.md:
```diff
@@ -0,0 +1,142 @@
+
+Here's how the `Scaffold` contributes to page design in FlutterFlow:
+
+**AppBar** : Scaffold allows you to easily include an `AppBar` at the top of the
+page,
+which can house the title, navigation controls, and other actions.
+
+**Floating Action Button (FAB)**: An action button that is commonly used for
+primary actions on the screen, like adding a new contact or composing a message.
+
+**Drawer & End-Drawer**: A slide-out menu for app navigation, accessible from
+the `AppBar` or by swiping from the side.
+
+**Body:** The main content area where you place your widgets. In FlutterFlow,
+you
+won't find a section explicitly labeled "Body". However, the widget tree that
+you assemble under your "PageNameWidget" essentially functions as the body of the
+page. This is where you add and organize all the interactive and static elements
+that make up the user interface of that particular page.
```
This should be a bulleted list

leighajarett commented on docs/resources/widgets/pages/pages.md:
```diff
@@ -0,0 +1,142 @@
+This organization is crucial for creating intuitive and user-friendly
+interfaces.
+
+#### Navigation
```
Each of these sections feel like a lot of whitespace. I think these can be bulleted lists?

leighajarett commented on docs/resources/widgets/pages/pages.md:
```diff
@@ -0,0 +1,142 @@
+feature of your app, whether it be a login screen, a profile page, or a settings
+menu.
+
+Here's how pages work in FlutterFlow:
```
This line doesnt seem necessary

leighajarett commented on docs/resources/widgets/pages/pages.md:
```diff
@@ -0,0 +1,142 @@
+you assemble under your "PageNameWidget" essentially functions as the body of the
+page. This is where you add and organize all the interactive and static elements
+that make up the user interface of that particular page.
+
+![body-page.png](..%2Fimgs%2Fbody-page.png)
+
```
I think it makes sense to show the page itself to actually show the user what an app bar, etc look like

leighajarett commented on docs/resources/widgets/pages/pages.md:
```diff
@@ -0,0 +1,142 @@
+
+Let's delve into the key moments in the page lifecycle:
+
+* **Initialization**: During initialization, you should set up
+  controllers or initialize the initial state of the page. This is where you may
+  set initial page states or fetch initial data from an API.
```
I think we should modify the css to add more spacing in between bullet points

leighajarett commented on docs/resources/widgets/pages/pages.md:
```diff
@@ -0,0 +1,142 @@
+
+To manage creation and updation of Page State variables, [see here](how-tos.md#page-state)
+
+## Page Lifecycle
```
A visualization would probably be helpful here

leighajarett commented on docs/resources/widgets/pages/pages.md:
```diff
@@ -0,0 +1,142 @@
+Let's delve into the key moments in the page lifecycle:
+
+* **Initialization**: During initialization, you should set up
+  controllers or initialize the initial state of the page. This is where you may
```
People may not know what controllers are. Can you explain this a bit more?

leighajarett commented on docs/resources/widgets/pages/pages.md:
```diff
@@ -0,0 +1,142 @@
+
+To manage creation and updation of Page State variables, [see here](how-tos.md#page-state)
+
+## Page Lifecycle
+
+In FlutterFlow and Flutter, understanding the page lifecycle is crucial as it
+involves the stages a page goes through from creation to disposal. Recognizing
+these stages is essential for managing resources and data effectively.
+
+Let's delve into the key moments in the page lifecycle:
+
```
I'm a bit confused as to what to make of this section. What do we want the user to take away from here? 

I think we probably want the user to understand when a page is reloaded (could be from the code or user interaction like refreshing a page) and what happens when its reloaded (i.e. page state variables are reset to their initial value) -- I feel like we're not providing enough helpful info here

leighajarett commented on docs/resources/widgets/pages/pages-configuration.md:
```diff
@@ -0,0 +1,129 @@
+sidebar_position: 1
+---
+#
+## Properties Panel
+
+In FlutterFlow, the properties panel located on the right side of the interface
+plays a crucial role in configuring and managing the various aspects of your
+pages. This panel is divided into several sections, each dedicated to different
+settings and adjustments you can make to customize your pages. Here’s a detailed
+look at what you can typically find and modify in this panel:
```
Are we going to talk about configuring widgets anywhere too?

I feel like we need property panel info to be somewhere more generic, not specific to pages

PoojaB26 replied:
Yep, added!

leighajarett commented on docs/resources/widgets/pages/pages-configuration.md:
```diff
@@ -0,0 +1,129 @@
+This section is where you set the fundamental aspects of a page’s layout and
+behavior, including:
+
+**Background Color:** This property allows you to set a background color for the
+entire page. You can choose a color that fits the theme and design of your app.
+
+**Safe Area:** When this toggle is enabled, the page content will be
+automatically adjusted to not overlap with the system status bar, navigation
+bar, and other critical device UI elements. This ensures that all elements of
+the page are visible and accessible on different devices.
+
+**Hide Keyboard on Tap:** Enabling this option makes the keyboard retract when
+the
+user taps anywhere outside the keyboard area on the screen. This is particularly
+useful for improving user experience by preventing the keyboard from obscuring
+content.
+
+**Disable Android Back Button:** This toggle, when activated, prevents the
+Android
+back button from affecting the navigation on this particular page. This can be
+useful in scenarios where you don't want users to navigate back to the previous
+screen easily, such as in a login or payment screen.
```
These can be bullets

leighajarett commented on docs/resources/widgets/pages/how-tos.md:
```diff
@@ -0,0 +1,481 @@
+toc_max_heading_level: 5
+---
+
+#
```
All pages should have an h1

leighajarett commented on docs/resources/widgets/pages/how-tos.md:
```diff
@@ -0,0 +1,481 @@
+
+#
+
+FlutterFlow allows you to easily create new pages using the Page Selector
+button, which is available in the Widget Tree menu (Tab+2). This will help you
+quickly start and add new pages to your app.
```
It feels a bit strange to me that were talking about creating an empty page after we've already shown the user how to configure the page

leighajarett replied:
Maybe we move the configurations to be a part of this how to guide

leighajarett commented on docs/resources/widgets/pages/how-tos.md:
```diff
@@ -0,0 +1,481 @@
+---
+title: How-to
```
I would give it a more descriptive name like "How to use pages"

leighajarett replied:
This page has a lot of information in it. But I worry we separated out the concepts and how-tos too much. It feels like we need to mesh them together more. I think we need to take another stab at organizing this information in a better way. 

For example, maybe having one page thats like "getting started with pages" where we describe what a page is / the scaffold thing, etc. and then show how to create a page, and configure it. 

Next page could be "navigating across pages" where we talk a bit about navigation and then the how-tos for navigation

... basically I think we give too much conceptual information and then all the how-tos here, it makes it hard to grasp the concepts and then apply them

leighajarett replied:
PLus I think there's just too much on this page, we need to break it up a bit

leighajarett commented on docs/resources/widgets/pages/how-tos.md:
```diff
@@ -0,0 +1,481 @@
+set "Disable Android Back Button" property on the destination page.
+:::
+
+**Replace Route:** Use this option to replace the current page in the navigation
+stack. For example, if a user navigates from Page A to Page B and then to Page
+C, pressing the back button on Page C would normally return to Page B. However,
+if **Replace Route** is enabled on Page B, the route changes to Page A -> Page
+C;
+therefore, pressing the back button on Page C will take the user back to Page A.
+
+**Transition Type:** This allows you to specify an animation that will be
+applied
+while navigating away from a screen. Options include **Default, Instant, Fade
+In,
+Slide Up, Slide Down, Slide Left, Slide Right,** and **Scale**.
+
+**Transition Duration:** Set the duration of the transition animation in
+milliseconds.
+
+**Page Parameters:** Use this to send data to the next page during navigation.
+
+![Nav.png](..%2Fimgs%2FNav.png)
```
I think these can be bullets

leighajarett commented on docs/resources/widgets/pages/how-tos.md:
```diff
@@ -0,0 +1,481 @@
+
+To utilize a template from FlutterFlow, follow these steps:
+
+<div style={{
+    position: 'relative',
+    paddingBottom: 'calc(56.67989417989418% + 41px)', // Keeps the aspect ratio and additional padding
+    height: 0,
+    width: '100%'
+}}>
```
There is some weird spacing after the arcades on this page. Seems like we need some padding or something

leighajarett reviewed (commented):
Added some comments - the content itself is looking really good. Most of my comments have to do with the order in which we present the content and tweaking how the pages look so its easier to make use of the content itself

leighajarett commented on docs/resources/widgets/pages/how-tos.md:
```diff
@@ -0,0 +1,481 @@
+</div>
+
+
+Other properties to consider include:
```
Feels like this can be a bulleted list

leighajarett reviewed (commented):
Left a bunch of comments. The cntent itself is great, but I dont love the order in which we display the content and how it's organized. Let me know if you wanna discuss over a call

PoojaB26 commented:
Hey @leighajarett, this is ready for review! Any missing pages or whole content, most probably is part of following PRs but you can mention it just to keep track of it!

leighajarett reviewed (changes requested):
This looks amazing! 

I think we could add some how-to / arcades to the widget configuration  - should we add those in this PR or follow up?

leighajarett reviewed (changes requested):
This looks amazing! 

I think we could add some how-to / arcades to the widget configuration  - should we add those in this PR or follow up?

PoojaB26 commented:
> This looks amazing!
> 
> I think we could add some how-to / arcades to the widget configuration - should we add those in this PR or follow up?

@leighajarett Let's do a follow up PR after this and merge this first? I can add a task to Linear. If its all good, please approve this PR then, thank you 🙌🏽
